src/coord/rmr/redise.c

```diff
@@ -13,60 +13,81 @@
-  size_t total_size = SlotRangeArray_SizeOf(num_ranges);
+  size_t total_size = SlotRangeArray_SizeOf(array_len(sh->slotRanges));
   RedisModuleSlotRangeArray* array = (RedisModuleSlotRangeArray*)rm_malloc(total_size);
-  array->num_ranges = num_ranges;
-  array->ranges[0].start = sh->startSlot;
-  array->ranges[0].end = sh->endSlot;
+  array->num_ranges = array_len(sh->slotRanges);
+  for (size_t i = 0; i < array_len(sh->slotRanges); i++) {
+    array->ranges[i] = sh->slotRanges[i];
+  }
   MRClusterShard csh = MR_NewClusterShard(&sh->node, array);
+  sh->node = (MRClusterNode){0}; // ownership transferred
   MRClusterTopology_AddShard(t, &csh);
 }
 
 /* Error replying macros, in attempt to make the code itself readable */
-#define ERROR_FMT(fmt, ...)                          \
-  ({                                                 \
-    char *err;                                       \
-    __ignore__(rm_asprintf(&err, fmt, __VA_ARGS__)); \
-    RedisModule_ReplyWithError(ctx, err);            \
-    rm_free(err);                                    \
-  })
+#define ERROR_FMT(fmt, ...) RedisModule_ReplyWithErrorFormat(ctx, fmt " at offset %zu", __VA_ARGS__, ac.offset)
 
 #define ERROR_BADVAL(arg, val) ERROR_FMT("Bad value for %s: %s", arg, val)
 #define ERROR_EXPECTED(exp, arg) ERROR_FMT("Expected " exp " but got `%s`", arg)
-#define ERROR_MISSING(arg) RedisModule_ReplyWithError(ctx, "Missing value for " arg)
+#define ERROR_MISSING(arg) ERROR_FMT("Missing value for %s", arg)
 
 #define ERROR_BAD_OR_MISSING(arg, ac_code)          \
   ({                                                \
     if (ac_code == AC_ERR_NOARG) {                  \
       ERROR_MISSING(arg);                           \
     } else {                                        \
-      ERROR_BADVAL(arg, AC_GetStringNC(&ac, NULL)); \
+      if (ac_code == AC_OK) ac.offset--;            \
+      const char *val = NULL;                       \
+      AC_GetString(&ac, &val, NULL, AC_F_NOADVANCE);\
+      ERROR_BADVAL(arg, val);                       \
+      if (ac_code == AC_OK) ac.offset++;            \
     }                                               \
   })
 
 #define VERIFY_ARG(arg)                                   \
   ({                                                      \
     if (!AC_AdvanceIfMatch(&ac, arg)) {                   \
-      const char *val = AC_GetStringNC(&ac, NULL);        \
-      ERROR_EXPECTED("`" arg "`", (val ? val : "(nil)")); \
+      const char *val = NULL;                             \
+      AC_GetString(&ac, &val, NULL, AC_F_NOADVANCE);      \
+      ERROR_EXPECTED("`" arg "`", (val ?: "(nil)"));      \
       goto error;                                         \
     }                                                     \
   })
 
 MRClusterTopology *RedisEnterprise_ParseTopology(RedisModuleCtx *ctx, RedisModuleString **argv,
                                                  int argc, uint32_t *my_shard_idx) {
```

Comment:
Consider adding the received topology to the log. I know that we had that in the past, and it was littering the logs sometimes, but in general, we should not receive that many `CLUSTER.SET` requests (unless there is an issue). Also, since we are changing it and it is more sensitive now, it may help with debugging issues (that hopefully will never come :) )